base/inc/AdePT/SparseArray.h

```diff
@@ -0,0 +1,417 @@
+
+  /** @brief Get current selection of indices */
+  __host__ __device__ __forceinline__ const Index_t *get_mask_array() const { return fMaskUsed; }
+
+  /** @brief Check if the element with index i is used */
+  __host__ __device__ __forceinline__ bool is_used(Index_t i) const
+  {
+    assert(i < N);
+    return (fMaskUsed[i >> 5].load() & (1 << (i & 31)));
+  }
+
+  /** @brief Check if container is fully distributed */
+  __host__ __device__ __forceinline__ bool full() const { return (size_shared() == N); }
+
+  __host__ __device__ __forceinline__ const_iterator begin() const { return const_iterator(&fData[0]); }
+
+  __host__ __device__ __forceinline__ const_iterator end() const { return const_iterator(&fData[fNused.load()]); }
+
+  __host__ __device__ __forceinline__ const_reference front() const { return *begin(); }
+
+  __host__ __device__ __forceinline__ const_reference back() const { return *(end() - 1); }
+
+  __host__ __device__ __forceinline__ const_pointer data() const { return &fData[0]; }
+
+  /** @brief Fills selection array fSelected with element indices passing the user predicate
+   *  @param pred Predicate function taking as arguments the SparseArray pointer and the element index
+   *  @returns Number of selected elements
+   */
+#ifdef COPCORE_CUDA_COMPILER
+  template <typename PREDICATE>
+  void select(PREDICATE pred_func)
+  {
+    int num_items = size_shared();
+    int num_warps = (num_items + 1023) >> 10; // a warp processes groups of 1024 elements
+    fNselected    = 0;
+    sa_detail::select_kernel<Array_t, PREDICATE><<<num_warps, 32>>>(this, pred_func, fSelected, &fNselected);
+  }
+
+  void select_used()
+  {
+    select([] __device__(int, const Array_t *) { return true; });
+  }
+#else
+  template <typename PREDICATE>
+  void select(PREDICATE pred_func)
+  {
+    // host version
+    const int num_items = size_shared();
+    fNselected          = 0;
+    for (int i = 0; i < num_items; ++i) {
+      if (is_used(i) && pred_func(i, this)) fSelected[fNselected++] = i;
+    }
+  }
+
+  void select_used()
+  {
+    // host version
+    select([](int, const Array_t *) { return true; });
+  }
+
+#endif
+
+  /** @brief Compacts used elements of this array into a destination array. */
+#ifdef COPCORE_CUDA_COMPILER
+  template <typename Array2_t>
+  void compact(Array2_t *dest)
+  {
+    static_assert(std::is_copy_constructible<Type>::value,
+                  "SparseArray::compact: Error: stored type is not copy constructible");
+    // select elements still in use
+    select_used();
+    assert(fNselected < (dest->size_max() - dest->size_shared()) &&
+           "SparseArray::compact: Error: not enough free space in destination array");
+    // copy to destination array
+    const int num_items = size_shared();
+    int num_blocks      = (num_items + 127) >> 7;
+    sa_detail::copy_kernel<Array_t, Array2_t><<<num_blocks, 127>>>(this, dest);
```

Comment:
I think you want to launch with `num_threads = 128`, with `127 = 128 - 1` to round up. Can you also change the computation of `num_blocks` to the more idiomatic `(num_items + num_threads - 1) / num_threads` (here and in other places)? The compiler will do the optimization to `>> 7` for you, but that way it's easier to change `num_threads`.